--- ./2.76/scripts/addons/io_mesh_stl/stl_utils.py~	2015-05-20 21:35:01.000000000 +0300
+++ ./2.76/scripts/addons/io_mesh_stl/stl_utils.py	2016-02-23 11:19:21.026078665 +0200
@@ -77,7 +77,7 @@
 
 def _header_version():
     import bpy
-    return "Exported from Blender-" + bpy.app.version_string
+    return "Created in touch-mapper.org. Exported from Blender-" + bpy.app.version_string
 
 
 def _is_ascii_file(data):
